fn/goroutine_manager.go

```diff
@@ -0,0 +1,77 @@
+package fn
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// ErrStopping is returned when trying to add a new goroutine while stopping.
+var ErrStopping = errors.New("can not add goroutine, stopping")
+
+// GoroutineManager is used to launch goroutines until context expires or the
+// manager is stopped. The Stop method blocks until all started goroutines stop.
+type GoroutineManager struct {
+	wg     sync.WaitGroup
+	mu     sync.Mutex
+	ctx    context.Context
+	cancel func()
+}
+
+// NewGoroutineManager constructs and returns a new instance of
```

Comment:
regarding the `GoroutineManager ` I mean it's good that we are introducing it, but I think we should disincentivize its use, because there is something wrong with the design if you need to use it, maybe we should mark this in the comment

Reply:
Fixed. Added the following note:

```
// NOTE: while it is safe to use GoroutineManager, it is recommended to change
// the design of the code using it to avoid launching goroutines except at start
// time. Ideally, goroutine launches and Stop() call should not compete.
```

Reply:
OK I have to chime in here. Why are we adding an abstraction that we are explicitly discouraging? Seems like this will be counterproductive in the long run.

Reply:
hmm ok, so you are actually fine with this way to use the waitGroup ? I thought it was a bandaid to the current situation to not rewrite the whole htlc switch/payment code, but using the waitgroup in that way is not good design imo, because the waitgroup is inherently thread safe and does not need a mutex generally when it is made sure Adds are called before Wait().

Reply:
My point is that *building infrastructure* to support *bad code design* is just weird. We should fix the design or patch over it. We don't want to make it *easier* to write bad code in the future.

Reply:
The common scenario is like this:

```
Some type 
  - New() returns new instance of the type
  - Some method starts an auxiliary goroutine which may keep working after the method completes
   (though this is not needed to have the problem with goroutines)
  - Method Stop() instructs all such background running workers to stop and waits for them to
   actually stop to avoid goroutine leak
```

(It is not only htlc swich, also found in ChannelRouter.SendPaymentAsync, maybe in other cases too.)

Such a scenario is not thread-safe to use a WaitGroup (without a mutex) to track goroutines.

But is this always a bad scenario which we should discourage in favor of a design using fixed number of goroutines started in `New()` and stopped in `Stop()`? The later scenario is thread safe for an unprotected WaitGroup.

Reply:
I don't think the issue is with fixed vs variable number of goroutines. This issue arises from having multiple threads responsible for managing the WaitGroup. By sending requests into some object's main event loop (which is responsible for managing the waitgroup) and having it launch the new goroutine would handily solve the issue. However right now we cannot make the assumption that the methods of an object are called from the same thread, which is what causes the problem.

This brings me to a new theorem of mine: WaitGroups that manage a runtime determined number of threads should only exist in the stack memory of a goroutine, *not* in a field on some object. If the number of goroutines is *statically known at compile time*, then it may be on the field of the object.

Reply:
I like the theorem! It is a good rule resulting in cleaner code. Also I think it is possible to enforce it at compile time (lint rule?).

Reply:
I don't know how to write lint rules right now but if that's possible that sounds great!

Reply:
I removed the note.